Add `KonnectCloudDataPlaneGroupConfig` admission policy

fgiudici opened the pull request:
**What this PR does / why we need it**:

Emit a warning when the `dataplane_groups.autoscale.type` is set to the deprecated `static` value.
Currently, deprecation warnings at field level require a validating admission policy
(see https://github.com/kubernetes/kubernetes/issues/131817).

**Which issue this PR fixes**

Fixes #1779 

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect significant changes

fgiudici committed:
Introduce `KonnectCloudDataPlaneGroupConfig` admission policy

Emit a warning when the dataplane_groups.autoscale.type is set to the
deprecated "static" value.
Currently, deprecation warnings at field level require a validating
admission policy
(see https://github.com/kubernetes/kubernetes/issues/131817).

Signed-off-by: Francesco Giudici <francesco.giudici@konghq.com>

pmalek commented on charts/kong-operator/templates/validation-policy-konnect.yaml:
```diff
@@ -0,0 +1,36 @@
+{{- if (semverCompare ">= 1.5.0" (include "kong.effectiveVersion" .)) }}
```
This is incorrect (leftover from porting the https://github.com/Kong/charts/tree/main/charts/gateway-operator).

This should be removed.

```suggestion
```

pmalek commented on charts/kong-operator/templates/validation-policy-konnect.yaml:
```diff
@@ -0,0 +1,36 @@
+  policyName: "autoscale.dataplanegroups.konnectclouddataplanegroupconfig.konnect.konghq.com"
+  validationActions: [Warn,Audit]
+{{- end -}}
+{{- end -}}
```
```suggestion
```

pmalek reviewed (changes requested):
We will need an option in `values.yaml` to control this so that users can deploy more than 1 helm release and do not get a conflict.

See related issue: https://github.com/Kong/kong-operator/issues/2154

fgiudici committed:
Update deprecation comment in the code.

Signed-off-by: Francesco Giudici <francesco.giudici@konghq.com>

pmalek reviewed (commented):
I'm wondering if we should rethink the CRD validation tests for VAPs.

Currently we have https://github.com/Kong/kong-operator/blob/85e364f9bcd34560feb9e966f77b14cc439e2c1e/test/crdsvalidation/dataplane_validatingadmissionpolicy_test.go with the ports VAP being installed with `kustomize` from https://github.com/Kong/kong-operator/blob/f2e367d6766b8653c2ade9195b489f40cb947cfb/config/default/validating_policies/kustomization.yaml.

I think we should unify everything to use the chart. WDYT? 

Bringing this up because I thought it would be worth to add a CRD validation test against the added VAP.

fgiudici committed:
validatingadmissionpolicy test: use resource from the chart directory

avoid to use the copy under the "config" dir and use the actual
admission policy available in the chart subdirectory.

This means moving out from kustomization to the rendering of the chart
resource and direct application of it.

Since a good amount of the kustomization package code was still valid,
moved the package name to "apply", keeping the
kustomization Apply() func (now called Kustomization()) and adding the
Template() one.

Signed-off-by: Francesco Giudici <francesco.giudici@konghq.com>

fgiudici commented:
> I'm wondering if we should rethink the CRD validation tests for VAPs.
> 
> Currently we have https://github.com/Kong/kong-operator/blob/85e364f9bcd34560feb9e966f77b14cc439e2c1e/test/crdsvalidation/dataplane_validatingadmissionpolicy_test.go with the ports VAP being installed with `kustomize` from https://github.com/Kong/kong-operator/blob/f2e367d6766b8653c2ade9195b489f40cb947cfb/config/default/validating_policies/kustomization.yaml.
> 
> I think we should unify everything to use the chart. WDYT?
> 
> Bringing this up because I thought it would be worth to add a CRD validation test against the added VAP.

Now the VAP tests use the templated resources from the chart.
Our testcase suite does not allow to check/catch warnings, so for the Konnect VAP added some code directly in the test to manage that.
The right way to address it is to extend our testcase suite with an "ExpectWarningMessage" functionality and then leverage that one: something to be addressed in a separate PR on kong/kubernetes-config (https://github.com/Kong/kubernetes-configuration/pull/602).

fgiudici committed:
Add test for konnect VAP

Signed-off-by: Francesco Giudici <francesco.giudici@konghq.com>

fgiudici committed:
make test.charts.golden.update

Signed-off-by: Francesco Giudici <francesco.giudici@konghq.com>

pmalek commented on test/helpers/apply/apply.go:
```diff
@@ -30,8 +32,44 @@ import (
-// Apply applies a kustomization to the cluster using the given rest.Config.
-func Apply(ctx context.Context, t *testing.T, cfg *rest.Config, dir string) {
+// Template applies templated resources to the cluster using the given rest.Config.
+func Template(t *testing.T, cfg *rest.Config, chartPath string, templates []string) {
```
I _think_ the change in these package is OK, going from tool oriented to verb oriented but I'm concerned that down the line when we want to add new actions we'd require a new package vs in the old package structure, we'd only need a new func. WDYT?

I believe though that in this context we'd want to rename this function and `templateRenderer` below so that they clearly mention they should be use with helm. WDYT?

fgiudici replied:
> I _think_ the change in these package is OK, going from tool oriented to verb oriented but I'm concerned that down the line when we want to add new actions we'd require a new package vs in the old package structure, we'd only need a new func. WDYT?
> 

So, the idea here was to render the template in order to use any template resource available from the chart.
I was aiming at a new package but then I realized that the Kustomization package had already all what was needed (apply(), readYaml(), buildDynamicResourceClient()) to apply the resource, only the template rendering was missing. 
Moreover, the "old" kustomize.Apply() function (renamed "Kustomization()") is no more used but it could be handy to be kept around.
This is why instead of creating a new package (and copying quite some code from kustomize) or renaming the kustomize one (and lose the kustomize.Apply()) I came up with this name-neutral package. 

I would be also in favor to split that change in 3 packages: the 3 common functions with the apply() exported (which is the one used by Kustomization and the Template one), keeping the kustomize package and adding a new one for the template rendering.
What would you prefer? (EDIT: pushed  a commit on top splitting in the 3 packages)

> I believe though that in this context we'd want to rename this function and `templateRenderer` below so that they clearly mention they should be use with helm. WDYT?

Something like helmRenderer() ? Ok (EDIT: now the function is helm.ApplyTemplate())

pmalek replied:
> I would be also in favor to split that change in 3 packages: the 3 common functions with the apply() exported (which is the one used by Kustomization and the Template one), keeping the kustomize package and adding a new one for the template rendering.
What would you prefer?

That could work 👍 Be wary of naming packages common or util though :) 

> Something like helmRenderer() ? Ok

This would idiomatically match an interface name in Go. I'd think more of something like `renderTemplate` (assuming the package is called `helm`)

pmalek commented on test/helpers/apply/apply.go:
```diff
@@ -30,8 +32,44 @@ import (
+func templateRenderer(t *testing.T, chartPath string, templates []string, helmArgs ...string) string {
+	t.Helper()
+	t2 := &testing.T{}
+	t2.Log("\nT2\n")
```
?

```suggestion
```

fgiudici replied:
🤦🏼  nice catch, a leftover

fgiudici committed:
Update test/helpers/apply/apply.go

Co-authored-by: Patryk Małek <patryk.malek@konghq.com>

fgiudici committed:
fixup! validatingadmissionpolicy test: use resource from the chart directory

Signed-off-by: Francesco Giudici <francesco.giudici@konghq.com>

pmalek commented on test/crdsvalidation/dataplane_validatingadmissionpolicy_test.go:
```diff
@@ -1,32 +1,104 @@
+	// ChartPath is the path relative to the project to the kong-operator chart
+	ChartPath = "charts/kong-operator"
+
+	// ValidationPolicy_ contain validation policies path relative to the kong-operator chart
```
Let's make this so that each `const` have its own go doc.

pmalek commented on test/crdsvalidation/dataplane_validatingadmissionpolicy_test.go:
```diff
@@ -1,32 +1,104 @@
+	ChartPath = "charts/kong-operator"
+
+	// ValidationPolicy_ contain validation policies path relative to the kong-operator chart
+	ValidationPolicyDataplane = "templates/validation-policy-dataplane.yaml"
```
This seems unused. I've checked out the repo on this branch and can't see any references to this

```suggestion
```

fgiudici replied:
Oh, right... I meant to use the const in the test but replicated the string.
Good catch! Now we use the const string in the test.

pmalek reviewed (commented):
Looks good, 2 minor nits, other than that we're ready to merge.

fgiudici committed:
fixup! validatingadmissionpolicy test: use resource from the chart directory

fgiudici committed:
Merge branch 'main' into issue1779
